core/src/main/java/com/linecorp/armeria/server/ContextPathServicesBuilder.java

```diff
@@ -0,0 +1,432 @@
+/*
+ * Copyright 2023 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import static com.linecorp.armeria.server.ServerBuilder.decorate;
+import static java.util.Objects.requireNonNull;
+
+import java.util.function.Function;
+
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.internal.server.RouteDecoratingService;
+import com.linecorp.armeria.internal.server.annotation.AnnotatedServiceExtensions;
+import com.linecorp.armeria.server.annotation.ExceptionHandlerFunction;
+import com.linecorp.armeria.server.annotation.RequestConverterFunction;
+import com.linecorp.armeria.server.annotation.ResponseConverterFunction;
+
+/**
+ * Builds {@link ServiceConfig}s for a {@link VirtualHostBuilder}. All {@link ServiceConfig}s
+ * built by this builder will be served under a context path.
+ *
+ * @param <T> the original type which will be returned once the {@link ServiceConfig}
+ * is built using {@link #and()}.
+ */
+final class ContextPathServicesBuilder<T> implements ServiceConfigsBuilder {
+
+    private final T parent;
+    private final VirtualHostBuilder virtualHostBuilder;
+
+    ContextPathServicesBuilder(T parent, VirtualHostBuilder virtualHostBuilder) {
+        this.parent = parent;
+        this.virtualHostBuilder = virtualHostBuilder;
+    }
+
+    @Override
+    public AbstractServiceBindingBuilder route() {
+        throw new UnsupportedOperationException("Not supported yet");
+    }
+
+    @Override
+    public AbstractBindingBuilder routeDecorator() {
+        throw new UnsupportedOperationException("Not supported yet");
+    }
```

Comment:
Note that these are not called from anywhere yet. These will be implemented when this class is exposed publicly in the next stage